### ISSUE TITLE
Support easyjson.Marshaler as field type

### ISSUE THREAD

**Issue by ReenExe** [mailru/easyjson#194]
I have struct
```golang
	type SocketCommand struct {
		CMD   string      `json:"cmd"`
		Value interface{} `json:"cmdval"`
	}
```

and auto generated Marshaler
```golang
func easyjsonCe9b818aEncodeEngine(out *jwriter.Writer, in SocketCommand) {
// ...
		if m, ok := in.Value.(easyjson.Marshaler); ok {
			m.MarshalEasyJSON(out)
		} else if m, ok := in.Value.(json.Marshaler); ok {
			out.Raw(m.MarshalJSON())
		} else {
			out.Raw(json.Marshal(in.Value))
		}
// ...
}
```

So i see some overhead `m, ok := in.Value.(easyjson.Marshaler); ok`, and think, how about change struct to another more optimize
```
	OptimizeSocketCommand struct {
		CMD   string             `json:"cmd"`
		Value easyjson.Marshaler `json:"cmdval"`
	}
```
after try
```bash
easyjson -all ./structs.go
```
i see:
```bash
interface type easyjson.Marshaler not supported: only interface{} is allowed
exit status 1
Bootstrap failed: exit status 1
```

But expect *success*

Proposal:
Add support for:
```
type Marshaler interface {
	MarshalEasyJSON(w *jwriter.Writer)
}

// Marshaler is an easyjson-compatible unmarshaler interface.
type Unmarshaler interface {
	UnmarshalEasyJSON(w *jlexer.Lexer)
}
```

and some combine type what cover Marshaler & Unmarshaler together.
See @121, but current issue is other, and expect
```golang
func easyjsonCe9b818aEncodeEngine(out *jwriter.Writer, in SocketCommand) {
// ...
			in.Value.MarshalEasyJSON(out)
// ...
}
```

**Comment by shmel1k** [comment]
Unfortunately, it is not possible now. When decoder/encoder sees that you have 'Unmarshaler' field in a struct, he considers it as an interface. If you just add a condition that allows to check interface type(in encoder/decoder), you will not be able to assign Marshaler/Unmarshaler in UnmarshalEasyJSON()/MarshalEasyJSON() generated functions so you get a compiler error. Thats why only empty interfaces are allowed.

Probably the best way to achieve that is to add some kind of
```go
type MarshalUnmarshaler interface{
    MarshalEasyJSON()
    UnmarshalEasyJSON()
}
```

and pass it to the struct you want, but easyjson must be patched for that anyway.

**Comment by ReenExe** [comment]
Если данное улучшения актуально то могу заняться и сделать pull request, @shmel1k что скажеш?

**Comment by shmel1k** [comment]
Ну, можно попробовать :)

**Comment by makarchuk** [comment]
This issue seemed to be inactive for a while, so I tried to figure it out. 
I think I did, but I'm not sure how to run tests for this project `go test` finds nothing and `go test ./...` fails on master. I'll be happy to run tests (and implement my own) ones I have a clue how to do so.

**Comment by shmel1k** [comment]
just run `make`. The binary file will be built and the tests will run.

**Comment by makarchuk** [comment]
Ok. Great. I'll come back with fixes soon enough (hopefully)

**Comment by shmel1k** [comment]
```go
type MarshalUnmarshaler interface {
        Marshal()
        Unmarshal()
}

type SomeStruct struct {
}

func (s *SomeStruct) Marshal() {

}

func (s *SomeStruct) Unmarshal() {

}

func (s *SomeStruct) DoSomething() {

}

func BenchmarkDirectCall(b *testing.B) {
        s := &SomeStruct{}
        for i := 0; i < b.N; i++ {
                s.Marshal()
        }
}

func BenchmarkIndirectCall(b *testing.B) {
        s := MarshalUnmarshaler(&SomeStruct{})
        for i := 0; i < b.N; i++ {
                if ss, ok := s.(MarshalUnmarshaler); ok {
                        ss.Marshal()
                }
        }
}
```

```goos: linux
goarch: amd64
BenchmarkDirectCall-8           2000000000               0.25 ns/op            0 B/op          0 allocs/op
BenchmarkIndirectCall-8         500000000                3.07 ns/op            0 B/op          0 allocs/op
```

**Comment by makarchuk** [comment]
It's not only about optimization in my case. I want to have some non-empty interface in my struct.

If this is implemented I can use any interface in my structs as long as it supports (un)marshalling.

**Comment by shmel1k** [comment]
https://play.golang.org/p/q1zOb118kIL

if really, your idea does not seem legit to me

**Comment by makarchuk** [comment]
Usecase that led me to this issue is the following:
I'm accepting a lot of JSON encoded structs and I'm forced to use some kind of `interface` in this structs, since some fields change type among those records, and I used `interface{}` which was `Ok` for my case. 

I used `easyjson` instead of regular `json` to improve preformance and it worked great.

Later I realized that as I'm working with this structures down the line I have no desire to constantly bother with type switches all the time. I can just use non empty `interface` (let's call it `MagicInterface`) for this fields and work with that.

Once I tried this `easyjson` refused to generate code for me, which is not reasonable, since my `MagicInterface` inculdes `easyjson.Unmarshaler` and `easyjson.Marshaler`.

I think it'd be reasonable for `easyjson` to check if type supports (un)marshaling already and if so, just use it.
